### PR TITLE
libusb: linux_usbfs: Fix parsing of descriptors for multi-configuration devices

### PR DESCRIPTION
**Pull request by httpstorm**
Backported an important fix for a regression in libusb-1.0.24, which caused multi-configuration devices not be recognised.
This fix restores the ability to tether iPhones.

Commit e2be556bd2 ("linux_usbfs: Parse config descriptors during device
initialization") introduced a regression for devices with multiple
configurations. The logic that verifies the reported length of the
configuration descriptors failed to count the length of the
configuration descriptor itself and would truncate the actual length by
9 bytes, leading to a parsing error for subsequent descriptors.

@neheb @flyn-org @nbd168 @aparcar @jow-

### CODE REVIEW

**Comment by adschm** [comment]
You are adding to patches here, while your commit message is only about one?

Apart from that, your commit title is (obviously) too long, and the Closes #xxx should be dropped. Depending on whether he authorized this submission, you might also need to remove Chris' Signed-off-by from the commit message (but _not_ the patches, where it should be kept.)

**Comment by neheb** [comment]
I don't think that version should be bumped.

**Comment by httpstorm** [comment]
Thank you, @adschm !
I shortened the title, removed Closes, and the additional signed-off. Added a note about the prerequisite patch.
I intentionally kept the original commits as two separate patches, so that it is easier to link them to the libusb repository.
Any feedback is welcome! I am learning with every contribution I make to this amazing project!

@neheb 
Hi Rosen!
I have kept the original  PKG_RELEASE=1. Let me know if any additional changes are required!

**Comment by neheb** [comment]
No I meant in the patches. PKG_RELEASE should be bumped. LIBUSB_NANO looks like some internal thing.

**Comment by httpstorm** [comment]
@neheb Hello Rosen!
PKG_RELEASE bummed to 2.
1. Would you like me to preserve both patches to be in sync with the first two commits after 1.0.24? Without any changes, the second patch will fail unless the first is also present.
2. Or would you rather prefer me use a single patch changing as little as possible?

**Comment by httpstorm** [comment]
@aparcar 
Hello Paul! I am sure it is a good thing to automate things, as long as it simplifies the maintenance and does not overcomplicate it. If I understand the article from your link correctly, then I should set PKG_RELEASE:=$(AUTORELEASE). That's easy. But then I should also remember to add a message to each commit, and I could not understand what exactly to add. So we replace a simple step with something complex. I would rather remember to increment the value manually as Rosen taught me. Of course if you tell me exactly what I should add to the commit message, I will add it.
I also need to meet the requirements for this PR to get accepted and move to fixing the perl package which does not build on macOS. I wish someone could help me bump perl to version 5.30.0 or newer which would resolve that.

**Comment by aparcar** [comment]
@httpstorm The suggested method reduces the manual changes required for package updates. There is no action from you required. If the maintainer decides to use the $(AUTORELEASE) variable, it's a set and forget. When you update a package version, we already require people to have a commit subject (not message body) which contains the package name followed by a short description.

**Comment by httpstorm** [comment]
@aparcar 
Sounds good. Thank you for clarifying, Paul! Let me know if any additional changes are needed.

- [x] PKG_RELEASE:=$(AUTORELEASE)

**Comment by httpstorm** [comment]
@neheb @flyn-org @nbd168 @aparcar @jow- @adschm 
Please let me know if I should make any changes!

**Comment by httpstorm** [comment]
ping @adschm @neheb @aparcar

**Review by neheb — APPROVED** [review]

File: package/libs/libusb/patches/001-Correct-a-typo-in-the-Changelog-and-clean-up-a-stray.patch [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+ delete mode 100644 test
+
+--- a/ChangeLog
++++ b/ChangeLog
+@@ -12,7 +12,7 @@ visit: http://log.libusb.info
+ * Darwin (macOS): use IOUSBDevice as darwin_device_class explicitly (#693)
+ * Linux: Drop support for kernel older than 2.6.32
+ * Linux: Provide an event thread name (#689)
+-* Linux: Wait until all USBs have been reaped before freeing them (#607)
++* Linux: Wait until all URBs have been reaped before freeing them (#607)
```
Comment (aparcar) [inline package/libs/libusb/patches/001-Correct-a-typo-in-the-Changelog-and-clean-up-a-stray.patch:21]: What is the change here?

**Comment by aparcar** [comment]
Merged, thank you.

**Review by httpstorm — COMMENTED** [review]
@aparcar 
Thank you for merging! Each of the two patches track a commit in the libusb history. I could have kept only patch 2 and adapted it, I thought it would be better to keep the upstream changes as they are.

**Comment by aparcar** [comment]
@adschm what do you think is the best approach there? This way it's closer to master.

**Comment by adschm** [comment]
I'm not sure what the exact question is?

**Comment by httpstorm** [comment]
@adschm 
1. Keep both patches with track two commits from upstream.
2. Remove patch 001, and remove the changes from version_nano.h in patch 002. This is the minimal fix.

**Comment by httpstorm** [comment]
@aparcar 
I just tested a clean build of the latest master on macOS, libusb works correctly. Thank you, Paul!

**Comment by adschm** [comment]
I did not follow this, so I'm still not sure what exactly the question would be. But if this is about editing backported upstream patches, it's prefer to not alter the upstream patches, but add additional patches for adjustments, so that the upstream patches can be removed during kernel update without additional thinking.

**Comment by httpstorm** [comment]
@adschm 
Yes, my patches, as well as the merge are copied from upstream commits. So they can be removed when we switch to the next version of libusb.

**Pull request closed without merging**